File: golang/2492.minimum-score-of-a-path-between-two-cities.go
import (
  "math"
)

type UF struct {
  cap int
  min []int
  par []int
}

func NewUF(cap int) *UF {
  par := make([]int, cap)
  minLen := make([]int, cap)

  for i := range cap {
    par[i] = i
    minLen[i] = math.MaxInt
  }

  return &UF{
    cap: cap,
    min: minLen,
    par: par,
  }
}

func (u *UF) find(x int) int {
  for x != u.par[x] {
    u.par[x] = u.par[u.par[x]]
    x = u.par[x]
  }

  return x
}

func (u *UF) join(x, y, w int) {
  parX, parY := u.find(x), u.find(y)

  minLen := min(u.min[parX], u.min[parY])
  u.par[parY] = parX
  u.min[parX] = min(w, minLen)
}

func (u *UF) getMinLen(x int) int {
  return u.min[u.find(x)]
}

func minScore(n int, roads [][]int) int {
    uf := NewUF(n)

    for _, v := range roads {
      uf.join(v[0] - 1, v[1] - 1, v[2])
    }

    return uf.getMinLen(n - 1)
}

File: rust/2492.minimum-score-of-a-path-between-two-cities.rs
struct UF {
  cap: usize,
  par: Vec<usize>,
  min: Vec<i32>,
}

impl UF {
  pub fn new(cap: usize) -> Self {

    return Self{
      cap: cap,
      par: Vec::from_iter(0..cap),
      min: vec![i32::MAX; cap],
    }
  }

  pub fn find(&mut self, mut x: usize) -> usize {
    while x != self.par[x] {
      self.par[x] = self.par[self.par[x]];
      x = self.par[x];
    }

    x
  }

  pub fn join(&mut self, x: usize, y: usize, mut w: i32) {
    let par_x = self.find(x);
    let par_y = self.find(y);

    w = self.min[par_x].min(w);
    w = self.min[par_y].min(w);

    self.par[par_y] = par_x;
    self.min[par_x] = w;
  }

  pub fn get_min(&mut self, x: usize) -> i32 {
    let par_x = self.find(x);

    self.min[par_x]
  }
}

impl Solution {
    pub fn min_score(n: i32, roads: Vec<Vec<i32>>) -> i32 {
        let mut uf = UF::new(n as usize);

        for r in roads.iter() {
          let from = r[0] as usize - 1;
          let to = r[1] as usize - 1;
          let len = r[2];

          uf.join(from, to, len);
        }

        uf.get_min(n as usize - 1)
    }
}